### PR TITLE
Standard and experiment docker

### PR DESCRIPTION
**Pull request by li126com**
## Motivation

Add standard and experiment dockerfiles for both ubuntu and centos as well as their READMEs and makefile.

## Modification

A new docker.Makefile in the root directory.
A new docker subdirectory containing Dockerfile-ubuntu and  Dockerfile-centos.
Two docker READMEs under the doc directory.
Add hyperlink to install.md.
A new experiment subdirectory containing Dockerfile-ubuntu ,  Dockerfile-centos, README-CN and README-EN.

## Use cases

When installing the environment, the user can directly use the standard docker image or build a image with the dockerfile.
We also provide dockerfiles and images for experiment of new flash-attention and torch.

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by sunpengsdu** [comment]
PR Message需要补充下

**Comment by sunpengsdu** [comment]
这个docker相关的代码和文档需要找专门的目录放，不能直接放根目录

File: Dockerfile-ubuntu [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ENV CC=${GCC_HOME}/bin/gcc
+ENV CXX=${GCC_HOME}/bin/c++
+
+
+FROM dep as intrenlm-dev
+COPY . /InternLM
+WORKDIR /InternLM
+ARG https_proxy
+ARG http_proxy
+ARG TORCH_CUDA_ARCH_LIST="3.7;5.0;6.0;7.0;7.5;8.0;8.6+PTX"
```
Comment (sunpengsdu) [inline Dockerfile-ubuntu:85]: 这个可以再看看，5.0 6.0之类的，显然有点低

File: docker.Makefile [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+DOCKER_IMAGE             ?= internlm
+DOCKER_FULL_NAME          = $(DOCKER_REGISTRY)/$(DOCKER_ORG)/$(DOCKER_IMAGE)
+
+CUDA_VERSION              = 11.7.1
+GCC_VERSION               = 10.2.0
+
+CUDNN_VERSION             = 8
+BASE_RUNTIME              =
+# ubuntu18.04  centos7
+BASE_OS                   = ubuntu18.04
```
Comment (sunpengsdu) [inline docker.Makefile:12]: 整个使用可以使用命令行传参指定BASE_OS

File: docker.Makefile [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+DOCKER_ORG               ?= my
+DOCKER_IMAGE             ?= internlm
+DOCKER_FULL_NAME          = $(DOCKER_REGISTRY)/$(DOCKER_ORG)/$(DOCKER_IMAGE)
+
+CUDA_VERSION              = 11.7.1
+GCC_VERSION               = 10.2.0
+
+CUDNN_VERSION             = 8
+BASE_RUNTIME              =
+# ubuntu18.04  centos7
```
[the remark this answers is not in the document]
Comment (li126com) [reply docker.Makefile:11]: 这个的话官方docker镜像就是centos7了没有再细化

File: Dockerfile-ubuntu (hunk 1 above)
Comment (li126com) [reply Dockerfile-ubuntu:85]: 跟guoteng商量了下"6.1;7.0;7.5;8.0;8.6+PTX"就支持6.1以上的主要GPU对应版本吧

File: docker.Makefile (hunk 2 above)
Comment (li126com) [reply docker.Makefile:12]: 可以直接命令行传参，我readme里加一下

**Comment by li126com** [comment]
> PR Message需要补充下

done

**Comment by li126com** [comment]
> 这个docker相关的代码和文档需要找专门的目录放，不能直接放根目录

done

**Comment by sunpengsdu** [comment]
ubuntu与centos的dockerfile，除了base OS不一样外，代码上还有哪些不一样？

**Comment by li126com** [comment]
> ubuntu与centos的dockerfile，除了base OS不一样外，代码上还有哪些不一样？

1. 基础包安装方式不同，ubuntu是apt，centos是yum安装
2. 基础安装包有所不同，ubuntu中的一些包在centos是不支持的需要安装其他的包或者用源码安装，centos也需要安装一些额外的包
3. 环境依赖安装过程有所不同，centos的环境变量默认值和ubuntu不太一样，具体的环境依赖安装过程也有所不同

**Comment by sunpengsdu** [comment]
一些源码安装的地方，看起来重复率挺高的，能想办法降低下重复的代码吗？

**Comment by li126com** [comment]
> 一些源码安装的地方，看起来重复率挺高的，能想办法降低下重复的代码吗？

不过有一个可以再优化一点的，就是不再需要用户手动传入dockerfile路径了，只要给出BASE_OS自动执行对应dockerfile

**Pull request merged**